src/rng.rs

```diff
@@ -393,14 +393,20 @@ impl Fill for [u8] {
     }
 }
 
-macro_rules! impl_fill {
+// This macro is unsafe to call: target types must support transmute from
+// random bits (i.e. all bit representations are valid).
+macro_rules! unsafe_impl_fill {
     () => {};
     ($t:ty) => {
         impl Fill for [$t] {
-            #[inline(never)] // in micro benchmarks, this improves performance
             fn fill<R: Rng + ?Sized>(&mut self, rng: &mut R) {
                 if self.len() > 0 {
-                    rng.fill_bytes(self.as_mut_bytes());
+                    rng.fill_bytes(unsafe {
+                        slice::from_raw_parts_mut(self.as_mut_ptr()
+                            as *mut u8,
+                            mem::size_of_val(self)
```

Comment:
I'll let @RalfJung chime in to confirm, but I believe this is unsound. In particular, you aren't allowed to interleave uses of raw pointers and references to the same object. You can fix this by moving the `mem::size_of_val(self)` above the call to `self.as_mut_ptr()`.

Reply:
Assigned the `size_of_val` to a local variable and added a tiny safety comment.